### PR TITLE
ci: introduce kail to collect cloudtest logs

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This resolves https://github.com/MaterializeInc/materialize/issues/16322.

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
The [build is still failing](https://buildkite.com/materialize/tests/builds/57020#01888f60-d4bc-437c-8288-39712eec2bb0), where can I find the logs of the docker install?

**Comment by philip-stoev** [comment]
@nrainer-materialize yes something is fishy, the ci-builder container did not get rebuilt even though you changed its Dockerfile. Let me try to debug.

File: ci/plugins/cloudtest/hooks/command [hunk 1]
```diff
@@ -27,6 +27,11 @@ fi
 
 date +"%Y-%m-%d %H:%M:%S" > step_start_timestamp
 
+ci_collapsed_heading "kail: Clean up running instances, start a new instance"
+run pkill -9 -f "^kail$" || true
+run rm -f "kail-output.log"
+run kail > "kail-output.log"
```
Comment (philip-stoev) [inline ci/plugins/cloudtest/hooks/command:33]: This needs to run inside the `ci-builder` container, but the `run` bash function does not provide that -- you need to use `bin/ci-builder run stable ...` for the purpose.

**Comment by philip-stoev** [comment]
> @nrainer-materialize yes something is fishy, the ci-builder container did not get rebuilt even though you changed its Dockerfile. Let me try to debug.

Ignore that, the container was properly rebuilt by the CI in your first commit.

File: ci/plugins/cloudtest/hooks/command (hunk 1 above)
Comment (nrainer-materialize) [reply ci/plugins/cloudtest/hooks/command]: Thank you!

File: ci/plugins/cloudtest/hooks/command [hunk 2]
```diff
@@ -39,5 +39,10 @@ ci_collapsed_heading "kind: Purging state from previous builds..."
 bin/ci-builder run stable test/cloudtest/reset
 rm -f kubectl-*.log
 
+ci_collapsed_heading "kail: Clean up running instances, start a new instance"
+bin/ci-builder run stable pkill -9 -f "^kail$" || true
+bin/ci-builder run stable rm -f "kail-output.log"
```
[the remark this answers is not in the document]
Comment (nrainer-materialize) [reply ci/plugins/cloudtest/hooks/command:44]: Ok, true. I will remove it

File: ci/plugins/cloudtest/hooks/command [hunk 3]
```diff
@@ -39,5 +39,10 @@ ci_collapsed_heading "kind: Purging state from previous builds..."
...
+bin/ci-builder run stable kail --context kind-cloudtest > "kail-output.log" &
```
[the remark this answers is not in the document]
Comment (nrainer-materialize) [reply ci/plugins/cloudtest/hooks/command:45]: It streams the logs of the different pods. Does this answer your question?

**Comment by philip-stoev** [comment]
@nrainer-materialize on top of your commit I applied this patch:

```
diff --git a/ci/plugins/cloudtest/hooks/command b/ci/plugins/cloudtest/hooks/command
index 7a66eb03e..9b4ba4518 100644
--- a/ci/plugins/cloudtest/hooks/command
+++ b/ci/plugins/cloudtest/hooks/command
@@ -40,7 +40,7 @@ bin/ci-builder run stable test/cloudtest/reset
 rm -f kubectl-*.log
 
 ci_collapsed_heading "kail: Start an instance"
-bin/ci-builder run stable --detach --name="kail" kail --context kind-cloudtest --log-level info --log-file "kail-output.log"
+bin/ci-builder run stable --detach --name="kail" kail --context kind-cloudtest --log-level info
 
 ci_uncollapsed_heading "cloudtest: Running \`bin/pytest ${run_args[*]}\`"
 bin/ci-builder run stable bin/pytest "${run_args[@]}" |& tee run.log
diff --git a/ci/plugins/cloudtest/hooks/post-command b/ci/plugins/cloudtest/hooks/post-command
index 6e609993b..3a713c11e 100755
--- a/ci/plugins/cloudtest/hooks/post-command
+++ b/ci/plugins/cloudtest/hooks/post-command
@@ -62,7 +62,8 @@ if [ -n "${CI_COVERAGE_ENABLED:-}" ]; then
 fi
 
 ci_unimportant_heading "kail: Stopping instance..."
-bin/ci-builder stop stable --name="kail"
+docker logs kail > kail-output.log 2>&1
+docker stop kail
```
and the kail output is now captured in the CI https://buildkite.com/materialize/tests/builds/57311#01889f1a-86fd-40c5-9fb5-8e0c4e65150b

I must admit I do not have 100% of an understanding as to why your approach did not work.

File: bin/ci-builder [hunk 4]
```diff
@@ -134,6 +134,15 @@ case "$cmd" in
         docker push materialize/ci-builder:"$cache_tag"
         ;;
     run)
+        if [[ $1 == "--detach" ]]; then
```
Comment (philip-stoev) [inline bin/ci-builder:137]: I think examples of more principled parsing using while + case can be found here:

https://stackoverflow.com/questions/192249/how-do-i-parse-command-line-arguments-in-bash

the answer has 3475 likes , wow

This way the order of `--detach` vs `--name` does not matter and they can be specified independently if needed.

File: ci/plugins/cloudtest/hooks/command [hunk 5]
```diff
@@ -39,5 +39,8 @@ ci_collapsed_heading "kind: Purging state from previous builds..."
 bin/ci-builder run stable test/cloudtest/reset
 rm -f kubectl-*.log
 
+ci_collapsed_heading "kail: Start an instance"
+bin/ci-builder run stable --detach --name="kail" kail --context kind-cloudtest --log-level info --log-file "kail-output.log"
```
Comment (philip-stoev) [inline ci/plugins/cloudtest/hooks/command:43]: make sure you do `docker kill kail || true` to clean up any existing kail instances.

Comment (nrainer-materialize) [reply ci/plugins/cloudtest/hooks/command:43]: I had that before but was wondering why I still need it. Don't I get a new EC2 instance? Or is it a used one?

Comment (philip-stoev) [reply ci/plugins/cloudtest/hooks/command:43]: Oh sorry, my fault, you do not need it, we already have a cleanup : 
```
ci_collapsed_heading ":docker: Purging all existing docker containers and volumes, regardless of origin"
docker ps --all --quiet | xargs --no-run-if-empty docker rm --force --volumes
```
earlier in the same file.

Buildkite keeps a cache of instances and your CI jobs may be given an instance that ran something previously that was not cleaned up completely.

**Review by philip-stoev — APPROVED** [review]

**Comment by nrainer-materialize** [comment]
> @nrainer-materialize on top of your commit I applied this patch:

Thanks for your help, @philip-stoev. It is finally working now. 😀

File: bin/ci-builder (hunk 4 above)
Comment (nrainer-materialize) [reply bin/ci-builder]: ✅

File: bin/ci-builder [hunk 6]
```diff
@@ -157,6 +180,15 @@ case "$cmd" in
             --env NIGHTLY_CANARY_CONFLUENT_CLOUD_API_SECRET
             --env PYPI_TOKEN
         )
+
+        if [[ $detach_container == "true" ]]; then
+            args+=("--detach")
+        fi
+
+        if [[ -n "$container_name_param" ]]; then
+            args+=("--name=$container_name_param")
```
Comment (def-) [inline bin/ci-builder:189]: I'm a bit confused why you had to handle the `--detach` and `--name` and then build them back into the same strings that they were before. Wouldn't the `$@` below take care of that as long as they are in the correct order?

File: ci/builder/README.md [hunk 7]
```diff
@@ -41,6 +41,9 @@ your latest changes. You can also run a command other than Bash directly:
 $ bin/ci-builder run stable echo "hello from abroad"
 ```
 
+You can specify the parameter `--detach` to start a detached Docker container.
+You can assign a name to the container using the parameter `--name`.
```
Comment (def-) [inline ci/builder/README.md:45]: One difference is that docker supports `--name=value`, but your script doesn't.

**Review by def- — APPROVED** [review]

File: bin/ci-builder (hunk 6 above)
Comment (nrainer-materialize) [reply bin/ci-builder:189]: `--detach` and `--name` need to run into `args`, which the rest needs to into `docker_command`. The latter ones need to come after the docker image definition (see line 267).

File: ci/builder/README.md (hunk 7 above)
Comment (nrainer-materialize) [reply ci/builder/README.md:45]: Happy to change it if that is considered more appropriate here (I actually had it like this before refactoring the arg parsing).

Comment (def-) [reply ci/builder/README.md:45]: I think it's fine as is.

**Pull request merged**